### ISSUE TITLE
About "norm_dim_in" in self.to_out

### ISSUE THREAD

**Issue by buyeah1109** [lucidrains/nGPT-pytorch#10]
Thanks for the great work. I notice that in the Attention and FFN, the output matrix (i.e., self.to_out) is normalized differently along the first dimension instead of the last dimension (normalizing along the last dimension is the default in your code, and this behavior is controlled by the flag "norm_dim_in"). 

I am wondering why the normalization is different for the output matrix. I was thinking that the author's goal of weight normalization is turning the computation of dot product to cosine similarity. But if we normalize along the first dimension of output matrix, then we are not calculating the cosine similarity between the intermediate state in FFN and weight vectors in output matrix. Correct me if I am wrong, really appreciate that.

**Comment by lucidrains** [comment]
@buyeah1109 hey Anthon! this is a good question which i don't really know the answer to

in the end, i went by this sentence and presumed embedding (model) dimension for all weight matrices

<img width="703" alt="Screen Shot 2024-10-28 at 6 27 26 AM" src="https://github.com/user-attachments/assets/8eb0a3b3-ab62-488c-ac8c-63d5d4b492a4">

however, there are papers that do allude that your way of doing it may work (look up the post-activation layernorm in feedforward in [normformer](https://arxiv.org/abs/2110.09456) paper, as well as the norm on the aggregated values coming out of attention, but before combining the heads, named [sub-ln](https://arxiv.org/abs/2210.06423))

**Comment by buyeah1109** [comment]
Thanks for your answer! That helps a lot. May I understand you are persuming the normalizing dimension is consistent across the entire model? Thus the normalizations in the code are all based on the "dim" dimension.

There are some good news. I observed a significant speed-up from nGPT in pretraining on the Openwebtext dataset. To achieve the same loss, nGPT requires much fewer tokens or training iterations than standard GPT2 baselines.

**Comment by lucidrains** [comment]
@buyeah1109 wow! you may be the first (although a lot of researchers keep their cards close) to report a positive result! (at least for the speedup)

how significant was it and was it holding all hyperparameters constant?

**Comment by lucidrains** [comment]
> Thanks for your answer! That helps a lot. May I understand you are persuming the normalizing dimension is consistent across the entire model? Thus the normalizations in the code are all based on the "dim" dimension.
> 
> There are some good news. I observed a significant speed-up from nGPT in pretraining on the Openwebtext dataset. To achieve the same loss, nGPT requires much fewer tokens or training iterations than standard GPT2 baselines.

yes that's what i'm assuming based on "all vectors forming the embedding dimension" phrase, but looking at some other papers (like the cosine sim network paper from 2017), it could be done the other way around when projecting out

in the end, you can only normalize one dimension of the weight matrix, and i can see both perspectives

**Comment by lucidrains** [comment]
@buyeah1109 maybe when i get some time i'll wire up the training script for openwebtext in this repo then... thanks for sharing your replication efforts

**Comment by lucidrains** [comment]
@buyeah1109 i could add some option to change up the normalizing dimension for those output projections in the `GPTExperimental` file for you to play around with, if you wish

**Comment by buyeah1109** [comment]
> @buyeah1109 wow! you may be the first (although a lot of researchers keep their cards close) to report a positive result! (at least for the speedup)
> 
> how significant was it and was it holding all hyperparameters constant?

Pretty good, it's about 3x faster than standard GPT2-124M. For hyperparameters, I just copy the default setting of standard GPT2 and tried different learning rates. I believe there will be a lot of space to improve by hyperparameter tuning.

**Comment by lucidrains** [comment]
@buyeah1109 so not without a ton of tuning.. that's great news.. 🙏 

looks like i'll put some more work into this next month. research isn't done here yet

**Comment by lucidrains** [comment]
now i'm wondering why i don't see anything on my toy dataset

i really need to move on from character level enwik8 to BPE on wikitext..

**Comment by lucidrains** [comment]
@buyeah1109 you didn't happen to train with mixed precision did you? also, your baseline was using rotary embeddings or no?

**Comment by buyeah1109** [comment]
> @buyeah1109 you didn't happen to train with mixed precision did you? also, your baseline was using rotary embeddings or no?

I trained with mixed percision. My GPU is Ampere so I think BF16 is supported and used in the mixed training. I used to train with V100s but ended up with NaN. For the baseline, i testify baseline both with and w/o the rotary embedding. RoPE slightly improved baseline' training losses.

**Comment by lucidrains** [comment]
@buyeah1109 wonderful! thank you!

**Comment by alxndrTL** [comment]
> Pretty good, it's about 3x faster than standard GPT2-124M. For hyperparameters, I just copy the default setting of standard GPT2 and tried different learning rates. I believe there will be a lot of space to improve by hyperparameter tuning.

Hey, would you care to share your baseline and nGPT code ? I tried to reproduce the results on the 124M scale but only got results comparable for nGPT and GPT.

**Comment by lucidrains** [comment]
@alxndrTL hey Alexandre, thanks for chiming in

i'm also seeing only "comparable" results, but then again, i've never been that great of an experimentalist, so will reserve judgement for a bit longer

**Comment by buyeah1109** [comment]
> > Pretty good, it's about 3x faster than standard GPT2-124M. For hyperparameters, I just copy the default setting of standard GPT2 and tried different learning rates. I believe there will be a lot of space to improve by hyperparameter tuning.
> 
> Hey, would you care to share your baseline and nGPT code ? I tried to reproduce the results on the 124M scale but only got results comparable for nGPT and GPT.

Sure! I used the code from nanoGPT to train the GPT2-124M baseline on openwebtext dataset. I use the default training configuration in nanoGPT for GPT2 except for the batchsize since I don't have 8x H100s lol. For nGPT, i keep using the nanoGPT training script and directly import model from this wonderful project and align the depth, width and dimension with GPT2. I didn't use QK-norm for nGPT.

**Comment by buyeah1109** [comment]
I may also share some data points. GPT2-124M achieves training loss around 3.04 after training with 16B tokens. nGPT achieves 2.94 after training with 16B tokens. nGPT achieves 3.0 after training with 8B tokens.

**Comment by buyeah1109** [comment]
I think your @alxndrTL  "comparable" results also matters. You may compare with my setting and find the difference, which could enlight us which part of the nGPT implementation is the most significant one for the training speed-up. It would help a lot!

**Comment by lucidrains** [comment]
@buyeah1109 thank you Anthon. between you, @jfpuget , and @inspirit , think i'll start paying attention to this

there was a separate camp of researchers who claim magnitude and even outliers are important. but who knows, in science you just have to be prepared to be proven wrong over and over again.

**Comment by alxndrTL** [comment]
Thanks for the details @buyeah1109 . Yes I will try again, what you saw gives me hope.

**Comment by inspirit** [comment]
In my experiments (different modalities including mel-spectorgrams and other embeddings, no text tokens) i noticed that setting `norm_dim_in=False` for input projection may render the whole model unusable, it still learns and reduce losses but predictions are far from acceptable. I ran few experiments changing only model first input projections to `norm_dim_in=True` and also setting everywhere in the model to always norm input dimension and it seems so far that first input projection very important in terms of which weight dimension to norm.

**Comment by lucidrains** [comment]
@inspirit yes that makes sense, as the magnitude in the continuous data will be lost without the network given some room to encode it as phase

i also got curious and spent the whole day yesterday trying out a normalized MLP for a small RL task, but while it learns and is stable, not really that much better than SOTA. while we are on this topic, i think i'll just also bring up the [xval paper](https://arxiv.org/abs/2310.02989), where it shows that a transformer can make use of magnitude on a token to generalize better for numerical tasks. play devil's advocate for this approach

**Comment by inspirit** [comment]
@lucidrains I wonder what you would recommend as an input projection for continuous data? especially if data dimensionality is 2-3x larger than transformer dim. I now have 2 ideas, first I have already verified and it works by setting norm_dim_in=True for input layer, second idea would be using just normal Linear layer without weight norm and just l2norm output of the layer. 
any thoughts?

**Comment by inspirit** [comment]
I guess another way is to use small MLP as input/output projection, the question is what weight norm options to use to project from data_dim to model_dim

**Comment by lucidrains** [comment]
@inspirit in the RL setting, what worked was just a linear followed by an activation, but if somehow heavily normalized networks like these take off, i'm sure people will be combing the lit for better ways to encode magnitude into phase. in other words, i don't know

**Comment by joonleesky** [comment]
I tried using a hyperspherical constraint in RL, and it worked great!
I ended up concatenating the constant to preserve the magnitude information!
https://dojeon-ai.github.io/SimbaV2/

**Comment by lucidrains** [comment]
@joonleesky wow 🚀 ! firstly, congratulations on Simba, it is amazing work! very excited to read more / try SimbaV2 once i get back to RL!

**Comment by lucidrains** [comment]
@joonleesky astounding results! 🙏

**Comment by joonleesky** [comment]
Thanks for the kind words! 😊 Let me know if you have any questions about the implementation. This nGPT repo helped me a lot with implementation!

**Comment by lucidrains** [comment]
@joonleesky will be trying it out on lunar lander as soon as i find the time! i'm really glad you were able to repurpose this architecture, as it would seem the original paper was busted by an error in the author's baseline 😞

**Comment by lucidrains** [comment]
@joonleesky are you a lion king fan?

**Comment by lucidrains** [comment]
@joonleesky did you see any discernible improvements with ppo? or you didn't try it?

**Comment by joonleesky** [comment]
I haven't tested this with PPO, but I’d expect the improvements to be relatively modest compared to SAC. The benefits are usually more noticeable when the model is prone to overfitting, particularly with a high UTD ratio (more updates per data collection).

That said, since PPO is on-policy, its UTD ratio is generally much lower than in off-policy setups. Still, increasing ppo_epochs (the number of updates per on-policy rollout) could lead to noticeable improvements.

**Comment by lucidrains** [comment]
@joonleesky thank you Hojoon

improving UTD on the off-policy front is def worthwhile (Sergey Levine would be pleased)

**Comment by lucidrains** [comment]
@joonleesky if you are ever in SF, let me know and i'll buy you dinner 😄
